### PR TITLE
 Detect stricter constraints on gats where clauses in impls vs trait

### PR DESCRIPTION
**Pull request by jackh726**
I might try to see if I can do a bit more to improve these diagnostics, but any initial feedback is appreciated. I can also do any additional work in a followup PR.

r? @estebank

### CODE REVIEW

**Review by estebank — COMMENTED** [review]
The code changes look reasonable, left some comments.

File: src/test/ui/generic-associated-types/impl_bounds.stderr [hunk 1]
```diff
@@ -5,24 +5,25 @@ LL |     type A<'a> where Self: 'static = (&'a ());
...
-note: lifetime parameter instantiated with the lifetime `'b` as defined on the associated item at 17:16
-  --> $DIR/impl_bounds.rs:17:16
-   |
-LL |     type B<'a, 'b> where 'b: 'a = (&'a(), &'b ());
-   |                ^^
-note: but lifetime parameter must outlive the lifetime `'a` as defined on the associated item at 17:12
-  --> $DIR/impl_bounds.rs:17:12
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found
+
+error[E0276]: impl has stricter requirements than trait
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/generic-associated-types/impl_bounds.stderr:19]: In that case we should turn one of these two into a `delay_as_bug()` to only emit one.

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/trait_impl_difference.rs [hunk 2]
```diff
@@ -11,44 +11,52 @@ use rustc_hir::def_id::DefId;
...
         }
+        if let RegionResolutionError::ConcreteFailure(origin, _, _) = error.clone() {
+            if let SubregionOrigin::CompareImplTypeObligation {
+                span,
+                item_name,
+                impl_item_def_id,
+                trait_item_def_id,
+            } = origin
+            {
+                self.emit_associated_type_err(span, item_name, impl_item_def_id, trait_item_def_id);
```
[the remark this answers is not in the document]
Comment (jackh726) [reply compiler/rustc_infer/src/infer/error_reporting/nice_region_error/trait_impl_difference.rs:57]: Yes, that fixed it.

File: src/test/ui/generic-associated-types/issue-86787.stderr [hunk 3]
```diff
@@ -9,7 +9,7 @@ LL | |     <Left as HasChildrenOf>::T: 'a,
 LL | |     <Right as HasChildrenOf>::T: 'a
    | |                                    - help: consider adding a where clause: `, <Left as HasChildrenOf>::T: 'a`
 LL | |     = Either<&'a Left::T, &'a Right::T>;
-   | |________________________________________^ ...so that the type `<Left as HasChildrenOf>::T` will meet its required lifetime bounds
+   | |________________________________________^ ...so that the definition in impl matches the definition from the trait
```
[the remark this answers is not in the document]
Comment (jackh726) [reply src/test/ui/generic-associated-types/issue-86787.stderr:12]: Maybe, but I'm planning to fix this diagnostic at some point anyways.

File: src/test/ui/generic-associated-types/impl_bounds.stderr (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackh726) [reply src/test/ui/generic-associated-types/impl_bounds.stderr]: Adding `return Some(ErrorReported)` fixed this

**Comment by jackh726** [comment]
@estebank I added `return Some(ErrorReported)` and that fixed the duplicate error. With that, I'm comfortable here. There are some GATs diagnostics that can be improved, but I'll continue to do that incrementally.

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/trait_impl_difference.rs [hunk 4]
```diff
@@ -11,44 +11,53 @@ use rustc_hir::def_id::DefId;
...
+            sup_origin,
+            _sup,
+        ) = error.clone()
+        {
+            if let (&Subtype(ref sup_trace), &Subtype(ref sub_trace)) = (&sup_origin, &sub_origin) {
+                if let (
+                    ValuePairs::Types(sub_expected_found),
+                    ValuePairs::Types(sup_expected_found),
+                    CompareImplMethodObligation { trait_item_def_id, .. },
+                ) = (&sub_trace.values, &sup_trace.values, &sub_trace.cause.code)
```
Comment (estebank) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/trait_impl_difference.rs:35]: We can remove one level of nesting by merging the second `if let` check with the first one.

File: src/test/ui/generic-associated-types/missing-where-clause-on-trait.stderr [hunk 5]
```diff
@@ -0,0 +1,11 @@
+error: `impl` associated type signature for `Assoc` doesn't match `trait` associated type signature
+  --> $DIR/missing-where-clause-on-trait.rs:9:5
+   |
+LL |     type Assoc<'a, 'b>;
+   |     ------------------- expected
+...
+LL |     type Assoc<'a, 'b> where 'a: 'b = ();
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found
```
Comment (estebank) [inline src/test/ui/generic-associated-types/missing-where-clause-on-trait.stderr:8]: For future work, I'd like us to point out the exact divergence (in this case the whole `where` clause).

**Review by estebank — APPROVED** [review]

**Comment by estebank** [comment]
@bors r+

If you want to address the `if let` "flattening", feel free to r=me again.

**Comment by estebank** [comment]
Failing tests:

```
    [debuginfo-cdb] debuginfo\basic-types.rs
    [debuginfo-cdb] debuginfo\msvc-pretty-enums.rs
    [debuginfo-cdb] debuginfo\pretty-std.rs
```

**Comment by jackh726** [comment]
weird

**Comment by jackh726** [comment]
I think spurious @bors retry

**Comment by ehuss** [comment]
@bors retry

Windows image update has broken things.

**Pull request merged**